traits/ctraits.c

```diff
@@ -1147,16 +1147,26 @@ _has_traits_trait ( has_traits_object * obj, PyObject * args ) {
         }
 
         dict = delegate->obj_dict;
-        if ( ((dict == NULL) ||
-              ((temp_delegate = (has_traits_object *) PyDict_GetItem( dict,
-                                          trait->delegate_name )) == NULL)) &&
-              ((temp_delegate = (has_traits_object *) has_traits_getattro(
-                  delegate, trait->delegate_name )) == NULL) )
-            break;
 
+        temp_delegate = NULL;
+        if (dict != NULL) {
+            temp_delegate = (has_traits_object *) PyDict_GetItem(
```

Comment:
Nope, that's the wrong thing to do:  `PyDict_GetItem` does *not* set an exception for a missing key; it merely returns `None`.